Very rough implementation of bcast for CuSparseVector

kshyatt opened the pull request:
VERY naive implementation here. Added some TODOs for obvious performance improvements I could make. Some issues I can't quite figure out yet (these are commented out in the new tests).

kshyatt committed:
Very rough implementation of bcast for CuSparseVector

maleadt commented on lib/cusparse/broadcast.jl:
```diff
@@ -433,21 +510,29 @@ function Broadcast.copy(bc::Broadcasted{<:Union{CuSparseVecStyle,CuSparseMatStyl
     # the kernels below parallelize across rows or cols, not elements, so it's unlikely
     # we'll launch many threads. to maximize utilization, parallelize across blocks first.
-    rows, cols = size(bc)
+    rows, cols = sparse_typ <: CuSparseVector ? (length(bc), 1) : size(bc)
```
`size(::Vector, 2)` returns 1; that could make this a little more idiomatic

maleadt commented on lib/cusparse/broadcast.jl:
```diff
@@ -295,16 +294,52 @@ function _getindex(arg::Union{CuSparseDeviceMatrixCSR,CuSparseDeviceMatrixCSC},
+    end
+    return
+end
+# TODO: unify CSC/CSR kernels
```
Isn't `compute_offsets_kernel` below already unified?

kshyatt replied:
Yes lol my rebase missed this comment

maleadt commented:
Cool! It's been a while since I looked at those kernels, so I only gave the changes a superficial look for now.

kshyatt commented:
Do you have any thoughts on my "pointer workspace" idea? I think this could really simplify things at the cost of additional allocations

maleadt commented:
> I think this could really simplify things at the cost of additional allocations

Can you elaborate what that would simplify?

How large would the allocations be? They're asynchronous, so as long as it doesn't scale in terms of the dense size I think that would be fine.

kshyatt commented:
Right now we're iterating through `iPtr` for every `CuSparseVector` argument in both the offset computation (which really just updates `total_nnz`) and during the actual broadcast computation. My idea is to use the offset computation to populate an array that's kind of like `CSRIterator`, which, for every row present in `output`, stores its location in `arg.iPtr` (if `arg` is a `CuSparseVector`) or `0` (if the row isn't present/`arg` is a scalar or dense).

kshyatt committed:
Uncomment broken tests

maleadt commented:
```
Reason: unsupported call to an unknown function (call to ijl_get_nth_field_checked)
Stacktrace:
 [1] getindex
   @ ./tuple.jl:31
 [2] iterate
   @ ./tuple.jl:72
 [3] compute_offsets_kernel
   @ /var/lib/buildkite-agent/builds/gpuci-6/julialang/cuda-dot-jl/lib/cusparse/broadcast.jl:329
```

vchuravy commented on lib/cusparse/broadcast.jl:
```diff
@@ -332,6 +367,30 @@ function compute_offsets_kernel(T::Type{<:Union{CuSparseMatrixCSR, CuSparseMatri
+
+function sparse_to_sparse_broadcast_kernel(f, first_row::Ti, last_row::Ti, output::CuSparseDeviceVector{Tv,Ti},
+                                           offsets::Union{AbstractVector,Nothing},
+                                           args::Vararg{Any, N}) where {Tv, Ti, N}
```
```suggestion
function sparse_to_sparse_broadcast_kernel(f::F, first_row::Ti, last_row::Ti, output::CuSparseDeviceVector{Tv,Ti},
                                           offsets::Union{AbstractVector,Nothing},
                                           args::Vararg{<:Any, N}) where {Tv, Ti, N, F}
```

kshyatt committed:
Use Vararg for compute_offsets_kernel

kshyatt committed:
Uncomment more tests and add a @. for matrix bcast

kshyatt committed:
Update lib/cusparse/broadcast.jl

Co-authored-by: Valentin Churavy <v.churavy@gmail.com>

kshyatt committed:
Fix test

kshyatt committed:
Fix test

Get tests passing with a hand unroll

Use offsets as intermediate ptr storage

Don't use atmoic

maleadt commented:
I can work around one of the issues by using generated functions and `@ntuple`:
```julia
@generated function sparse_to_dense_broadcast_kernel(::Type{<:CuSparseVector}, f,
                                                     output::CuDeviceArray{Tv},
                                                     args...) where {Tv}
    N = length(args)
    quote
        # every thread processes an entire row
        row = threadIdx().x + (blockIdx().x - 1i32) * blockDim().x
        row > length(output) && return

        # set the values for this row
        vals = @ntuple $N i -> begin
            arg = @inbounds args[i]
            _getindex(arg, row, 0)::Tv
        end
        out_val = f(vals...)
        @inbounds output[row] = out_val
        return
    end
end
```

... but other issues persist. It's curious that this only happens with CuSparseVector, using code patterns (`args...`, untyped `f`, etc) that work perfectly fine with sparse matrices. Maybe because of the complexity of `CuSparseVector`'s `_getindex`? Forcing it to inline doesn't help, in any case.

maleadt committed:
Work around bad codegen.

kshyatt commented:
> Maybe because of the complexity of `CuSparseVector`'s `_getindex`

But then why does it work fine if you only have sparse vectors in the broadcast? The failures only start when you add in a dense vector...

maleadt commented:
Yeah not sure. Heterogeneity? The typed IR looks perfectly fine, but in the generated LLVM you can see that the closure passed to `ntuple` is not inlined (and presumably insufficiently optimized) anymore, hence the `@inline` helping in that case.

kshyatt commented:
I wonder, if we made the `vals = ntuple(...)` thing a separate function, might that help? I'll give it a whirl.

maleadt commented:
I think the `@inline` is a nice fix/workaround though. Similarly how I prefer the `@generated` + `@nany`, not to have to manually outline functionality that isn't actually reusable.

kshyatt commented:
OK, made `offsets` a longer-lived iterator store to hold the locations in `iPtr`/`nzVal` - in theory, this kind of thing should allow us to also do broadcasts of sparse vectors into sparse matrices and stuff.

maleadt commented on lib/cusparse/broadcast.jl:
```diff
@@ -1,5 +1,7 @@
 using Base.Broadcast: Broadcasted
 
+using Base.Cartesian: @nany
```
Unused now?

maleadt commented on lib/cusparse/broadcast.jl:
```diff
@@ -288,15 +289,18 @@ end
+@inline function _getindex(arg::CuDeviceArray{Tv}, I, ptr)::Tv where {Tv}
+    return @inbounds arg[I]::Tv
+end
+@inline _getindex(arg, I, ptr) = Broadcast._broadcast_getindex(arg, I)
```
Do these `@inline`s help with compilation issues?

maleadt replied:
Also why is the third arg `ptr` here but `is_nnz` in call sites? Can you add comments on some of the behavior here (e.g. the exact preprocessing you're doing in some of the kernels)?

kshyatt replied:
Yes can do

maleadt commented on test/Project.toml:
```diff
@@ -5,6 +5,7 @@ BFloat16s = "ab4f0b2a-ad5b-11e8-123f-65d77653426b"
 CUDA_Driver_jll = "4ee394cb-3365-5eb0-8335-949819d2adfc"
 CUDA_Runtime_jll = "76a88914-d11a-5bdc-97e0-2f5a05c973a2"
 ChainRulesCore = "d360d2e6-b24c-11e9-a2a3-2a2ae2dbcce4"
+Cthulhu = "f68482b8-f384-11e8-15f7-abe071a5a75f"
```
Not needed?

kshyatt replied:
Yeah whoops!

kshyatt committed:
Cleanup

kshyatt committed:
Remove unneeded using

maleadt committed:
Clean-ups.
